src/sentry/api/serializers/models/project.py

```diff
@@ -283,33 +284,44 @@ def get_attrs(self, item_list, user):
             environments_by_project[project_env['project_id']].append(
                 project_env['environment__name'])
 
-        release_project_envs = list(ReleaseProjectEnvironment.objects.filter(
-            project__in=item_list,
-            last_deploy_id__isnull=False
-        ).values('release__version', 'environment__name', 'last_deploy_id', 'project__id'))
-
-        deploys = dict(
-            Deploy.objects.filter(
-                id__in=[
-                    rpe['last_deploy_id'] for rpe in release_project_envs]).values_list(
-                'id',
-                'date_finished'))
-
+        cursor = connection.cursor()
+        cursor.execute(
+            """
+            select srpe.project_id, se.name, sr.version, date_finished
+            -- Finally, filter to the top row each each project/environment and get the environment name and release version.
+            from (
+                -- Next we join to deploys and rank based recency of latest deploy for each project/environment.
+                select srpe.project_id, srpe.release_id, srpe.environment_id, sd.date_finished,
+                row_number() OVER (partition by (srpe.project_id, srpe.environment_id) order by sd.date_finished desc) row_num
+                from
+                (
+                    -- First we fetch all related ReleaseProjectEnvironments, then filter to the x most recent for
+                    -- each project/environment that actually have a deploy. This cuts out a lot of data volume
+                    select *
+                    from (
+                        select *, row_number() OVER (partition by (srpe.project_id, srpe.environment_id) order by srpe.id desc) row_num
+                        from sentry_releaseprojectenvironment srpe
+                        where srpe.last_deploy_id is not null
+                        and project_id = ANY(%s)
+                    ) srpe
+                    where row_num <= %s
```

Comment:
Do many users have more than 10 environments for a project?

Reply:
Not sure, but it doesn't matter if they do. This will give us 10 rows per project/environment, not 10 rows total like a limit would.